src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -661,65 +661,63 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
     })
   }
 
-  def groupVariantsBy(keyExpr: String, aggExpr: String, partitionKey: java.util.ArrayList[String]): MatrixTable =
-    groupVariantsBy(keyExpr, aggExpr, Option(partitionKey).map(_.asScala.toArray))
+  def aggregateRowsAlongKey(aggExpr: String): MatrixTable = {
 
-  def groupVariantsBy(keyExpr: String, aggExpr: String, partitionKey: Option[Array[String]] = None): MatrixTable = {
-    val fullRowType = rvRowType
-    val vEC = EvalContext(Map(Annotation.GLOBAL_HEAD -> (0, globalType),
-      "va" -> (1, rowType)))
-    vEC.set(0, globals)
-    val (keyNames, keyTypes, keyFs) = Parser.parseNamedExprs(keyExpr, vEC)
-    partitionKey.foreach(pk => assert(keyNames.startsWith(pk)))
-    partitionKey.foreach(pk => assert(pk.nonEmpty))
-
-    val keyStruct = TStruct(keyNames.zip(keyTypes): _*)
-
-    val SampleFunctions(zero, seqOp, combOp, resultOp, newEntryType) = Aggregators.makeSampleFunctions(this, aggExpr)
-    val pk: IndexedSeq[String] = partitionKey.map(_.toFastIndexedSeq).getOrElse(keyNames.toFastIndexedSeq)
+    val SampleFunctions(zero, seqOp, resultOp, newEntryType) = Aggregators.makeSampleFunctions(this, aggExpr)
+    val newRowType = matrixType.orvdType.kType
     val newMatrixType = MatrixType.fromParts(globalType, colKey, colType,
-      pk, keyNames, keyStruct, newEntryType)
+      rowPartitionKey, rowKey, newRowType, newEntryType)
 
-    val localRowType = rowType
-    val localEntriesIndex = entriesIndex
-    val keyedRDD = rvd.rdd.mapPartitions { it =>
-      val fullRow = new UnsafeRow(fullRowType)
-      it.map { rv =>
-        fullRow.set(rv)
-        val row = fullRow.deleteField(localEntriesIndex)
-        vEC.set(1, row)
-        val k = Annotation.copy(keyStruct, Row.fromSeq(keyFs()))
-        k -> rv
-      }
-    }
-
-    // FIXME: this shuffles twice
     val newRVType = newMatrixType.rvRowType
-    val rdd = keyedRDD
-      .aggregateByKey(zero)(seqOp, combOp)
-      .mapPartitions { it =>
+    val localRVType = rvRowType
+    val selectIdx = matrixType.orvdType.kRowFieldIdx
+    val keyOrd = matrixType.orvdType.kRowOrd
+
+    val newRVD = rvd.mapPartitionsPreservesPartitioning(newMatrixType.orvdType) { it =>
```

Comment:
a note that the old model benefited from map-side combine resulting in a smaller shuffle. We'll optimize this later.

Reply:
(no action necessary)